### ISSUE TITLE
Classifier not training in scikit-learn < 1.2.2 

### ISSUE THREAD

**Issue by jluethi** [fractal-napari-plugins-collection/napari-feature-classifier#27]
Classifier training is not working in napari 0.4.17 (at least on some data where it's working in 0.5.0). When running it, I get `INFO: Not enough annotations made. Please make more annotations.`

This is a consequence of the try-catch for ValueErrors during classifier training. When not catching this ValueError, we get the following error message (not the one the try catch was designed for => maybe we should make this less broad?)

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File ~/mambaforge/envs/classifier-dev/lib/python3.9/site-packages/psygnal/_signal.py:950, in SignalInstance._run_emit_loop(self=<SignalInstance 'changed' on PushButton(value=False, annotation=None, name='')>, args=(False,))
    949 try:
--> 950     caller.cb(args)
        caller = <psygnal._weak_callback._WeakMethod object at 0x2b16d7400>
        args = (False,)
    951 except Exception as e:

File ~/mambaforge/envs/classifier-dev/lib/python3.9/site-packages/psygnal/_weak_callback.py:341, in _WeakMethod.cb(self=<psygnal._weak_callback._WeakMethod object>, args=())
    340     args = args[: self._max_args]
--> 341 func(obj, *self._args, *args, **self._kwargs)
        obj = <Container ()>
        func = <function ClassifierRunContainer.run at 0x2b34ae9d0>
        args = ()
        self = <psygnal._weak_callback._WeakMethod object at 0x2b16d7400>
        self._args = ()
        self._kwargs = {}

File ~/Dropbox/Joel/FMI/Code/napari-feature-classifier/src/napari_feature_classifier/classifier_widget.py:299, in ClassifierRunContainer.run(self=<Container ()>)
    298 self.add_features_to_classifier()
--> 299 self._classifier.train()
        self = <Container ()>
        self._classifier = Classifier
                                 annotations   area      hash
roi_id                    label                              
r04c07f02p01-ch1sk1fk1fl1 89               1  198.0  0.356967
                          136              1  174.0  0.277897
                          202              1  192.0  0.437402
                          258              1  207.0  0.292852
                          362              1  130.0  0.015523
                          405              1  173.0  0.559590
                          546              1  127.0  0.258025
                          562              1  166.0  0.174446
                          693              1  176.0  0.737419
                          697              1  150.0  0.422359
                          982              1  204.0  0.170698
                          1114             2   63.0  0.292251
                          1393             2   32.0  0.276789
                          1566             2   49.0  0.584304
                          1579             2   43.0  0.210398
                          1628             2   30.0  0.980660
                          1717             2   47.0  0.913480
                          1737             2   27.0  0.765238
    300 try:

File ~/Dropbox/Joel/FMI/Code/napari-feature-classifier/src/napari_feature_classifier/classifier.py:96, in Classifier.train(self=Classifier
                                 anno...              1737             2   27.0  0.765238)
     94 self._classifier.fit(X_train, y_train)
---> 96 f1 = f1_score(y_test, self._classifier.predict(X_test), average="macro")
        X_test =                                  area
roi_id                    label      
r04c07f02p01-ch1sk1fk1fl1 1628   30.0
                          1717   47.0
        y_test = roi_id                     label
r04c07f02p01-ch1sk1fk1fl1  1628     2
                           1717     2
Name: annotations, dtype: Int64
        self = Classifier
                                 annotations   area      hash
roi_id                    label                              
r04c07f02p01-ch1sk1fk1fl1 89               1  198.0  0.356967
                          136              1  174.0  0.277897
                          202              1  192.0  0.437402
                          258              1  207.0  0.292852
                          362              1  130.0  0.015523
                          405              1  173.0  0.559590
                          546              1  127.0  0.258025
                          562              1  166.0  0.174446
                          693              1  176.0  0.737419
                          697              1  150.0  0.422359
                          982              1  204.0  0.170698
                          1114             2   63.0  0.292251
                          1393             2   32.0  0.276789
                          1566             2   49.0  0.584304
                          1579             2   43.0  0.210398
                          1628             2   30.0  0.980660
                          1717             2   47.0  0.913480
                          1737             2   27.0  0.765238
        self._classifier = RandomForestClassifier()
     97 # napari_info("F1 score on test set: {}".format(f1))

File ~/mambaforge/envs/classifier-dev/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1146, in f1_score(y_true=roi_id                     label
r04c07f02p01-ch...       1717     2
Name: annotations, dtype: Int64, y_pred=<class 'numpy.ndarray'> (2,) float64, labels=None, pos_label=1, average='macro', sample_weight=None, zero_division='warn')
   1021 """Compute the F1 score, also known as balanced F-score or F-measure.
   1022 
   1023 The F1 score can be interpreted as a harmonic mean of the precision and
   (...)
   1144 array([0.66666667, 1.        , 0.66666667])
   1145 """
-> 1146 return fbeta_score(
        y_true = roi_id                     label
r04c07f02p01-ch1sk1fk1fl1  1628     2
                           1717     2
Name: annotations, dtype: Int64
        y_pred = <class 'numpy.ndarray'> (2,) float64
        labels = None
        pos_label = 1
        average = 'macro'
        sample_weight = None
        zero_division = 'warn'
   1147     y_true,
   1148     y_pred,
   1149     beta=1,
   1150     labels=labels,
   1151     pos_label=pos_label,
   1152     average=average,
   1153     sample_weight=sample_weight,
   1154     zero_division=zero_division,
   1155 )

File ~/mambaforge/envs/classifier-dev/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1287, in fbeta_score(y_true=roi_id                     label
r04c07f02p01-ch...       1717     2
Name: annotations, dtype: Int64, y_pred=<class 'numpy.ndarray'> (2,) float64, beta=1, labels=None, pos_label=1, average='macro', sample_weight=None, zero_division='warn')
   1169 """Compute the F-beta score.
   1170 
   1171 The F-beta score is the weighted harmonic mean of precision and recall,
   (...)
   1284 array([0.71..., 0.        , 0.        ])
   1285 """
-> 1287 _, _, f, _ = precision_recall_fscore_support(
        y_true = roi_id                     label
r04c07f02p01-ch1sk1fk1fl1  1628     2
                           1717     2
Name: annotations, dtype: Int64
        y_pred = <class 'numpy.ndarray'> (2,) float64
        beta = 1
        labels = None
        pos_label = 1
        average = 'macro'
        sample_weight = None
        zero_division = 'warn'
   1288     y_true,
   1289     y_pred,
   1290     beta=beta,
   1291     labels=labels,
   1292     pos_label=pos_label,
   1293     average=average,
   1294     warn_for=("f-score",),
   1295     sample_weight=sample_weight,
   1296     zero_division=zero_division,
   1297 )
   1298 return f

File ~/mambaforge/envs/classifier-dev/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1573, in precision_recall_fscore_support(y_true=roi_id                     label
r04c07f02p01-ch...       1717     2
Name: annotations, dtype: Int64, y_pred=<class 'numpy.ndarray'> (2,) float64, beta=1, labels=None, pos_label=1, average='macro', warn_for=('f-score',), sample_weight=None, zero_division='warn')
   1572     raise ValueError("beta should be >=0 in the F-beta score")
-> 1573 labels = _check_set_wise_labels(y_true, y_pred, average, labels, pos_label)
        labels = None
        y_true = roi_id                     label
r04c07f02p01-ch1sk1fk1fl1  1628     2
                           1717     2
Name: annotations, dtype: Int64
        y_pred = <class 'numpy.ndarray'> (2,) float64
        average = 'macro'
        pos_label = 1
   1575 # Calculate tp_sum, pred_sum, true_sum ###

File ~/mambaforge/envs/classifier-dev/lib/python3.9/site-packages/sklearn/metrics/_classification.py:1374, in _check_set_wise_labels(y_true=roi_id                     label
r04c07f02p01-ch...       1717     2
Name: annotations, dtype: Int64, y_pred=<class 'numpy.ndarray'> (2,) float64, average='macro', labels=None, pos_label=1)
   1372     raise ValueError("average has to be one of " + str(average_options))
-> 1374 y_type, y_true, y_pred = _check_targets(y_true, y_pred)
        y_true = roi_id                     label
r04c07f02p01-ch1sk1fk1fl1  1628     2
                           1717     2
Name: annotations, dtype: Int64
        y_pred = <class 'numpy.ndarray'> (2,) float64
   1375 # Convert to Python primitive type to avoid NumPy type / Python str
   1376 # comparison. See https://github.com/numpy/numpy/issues/6784

File ~/mambaforge/envs/classifier-dev/lib/python3.9/site-packages/sklearn/metrics/_classification.py:95, in _check_targets(y_true=roi_id                     label
r04c07f02p01-ch...       1717     2
Name: annotations, dtype: Int64, y_pred=<class 'numpy.ndarray'> (2,) float64)
     94 if len(y_type) > 1:
---> 95     raise ValueError(
        type_true = 'unknown'
        type_pred = 'binary'
     96         "Classification metrics can't handle a mix of {0} and {1} targets".format(
     97             type_true, type_pred
     98         )
     99     )
    101 # We can't have more than one value on y_type => The set is no more needed

ValueError: Classification metrics can't handle a mix of unknown and binary targets
```

Not sure why we're running into a `ValueError: Classification metrics can't handle a mix of unknown and binary targets` with this here, will need to dig a bit

Thanks @fdsteffen for pointing us to this!

Next steps:
- [x] Reproduce with the minimal test data in 0.4.17
- [x] Create a test case that fails for this with the 0.4.17 napari (which is also the one used during automated testing)
- [x] Find the source of this issue & fix it

**Comment by jluethi** [comment]
Can confirm that it is also broken in our minimal example using the little `dev_main.py` script => seems to be a larger issue with the 0.1.0 plugin and the 0.4.17 napari release.

**Comment by jluethi** [comment]
I don't understand why the `test_running_classification_through_widget` test in `test_classifier_widget` is passing. It's asserting that all the predictions are not NaN and they would be if the training fails, I think. Investigating...

**Comment by jluethi** [comment]
And the story continues to get weirder. Locally, `test_running_classification_through_widget` actually fails in my old-ish napari env. The NaN assertion correctly catches this, as it was designed to do!

I'll rebuild a fresh napari env, just installing those dependencies and try again. Because those same tests are actually passing in our automated testing on Github (=> we should be catching this issue!)

**Comment by jluethi** [comment]
The issue is not actually with the napari version, I tried with 0.4.17 in a fresh environment and there, everything worked fine. 
I next tested the scikit-learn version. I had 1.2.2 in the newer environment where the tests passed, but only 1.2.1 in the old one. Downgrading to 1.2.1 raises this issue again and makes the tests fail. No idea why we depend on the 1.2.2 version of scikit-image.
@MaksHess Any idea? I'll prepare a PR to force scikit-image >= 1.2.2 so we can just fix it that way, unless you see an obvious fix we could do to work in older scikit-image versions (but not worth investing a lot of time in, maybe their bug or how they handle NaNs or something?
@fdsteffen Can you confirm that you also had scikit-image < 1.2.2 in your environment?

**Comment by fdsteffen** [comment]
Yes, indeed I was running on scikit-learn 1.2.1 in my local environment. Many thanks for digging into this @jluethi!

**Comment by fdsteffen** [comment]
Just to quickly follow up, I guess this is indeed related to the issue raised in scikit-learn/scikit-learn#25578 and fixed with patch release 1.2.2 (scikit-learn/scikit-learn#25638 and scikit-learn/scikit-learn#25692). Updating the dependencies is likely the fastest way forward.

**Comment by jluethi** [comment]
@fdsteffen These issues are now fixed in the 0.1.1 release that has just been deployed to pypi. Thanks again for reporting them and feel free to open more issues if you run into any other bugs or gaps in our documentation!
